Enhance `needless_borrow` to consider trait implementations

smoelius opened the pull request:
The proposed enhancement causes `needless_borrow` to suggest removing `&` from `&e` when `&e` is an argument position requiring trait implementations, and `e` implements the required traits. Example:
```
error: the borrowed expression implements the required traits
  --> $DIR/needless_borrow.rs:131:51
   |
LL |     let _ = std::process::Command::new("ls").args(&["-a", "-l"]).status().unwrap();
   |                                                   ^^^^^^^^^^^^^ help: change this to: `["-a", "-l"]`
```


r? @Jarcho

changelog: Enhance `needless_borrow` to consider trait implementations

Jarcho commented:
Probably won't have time to look at this for a few days. Does this handle something like:
```rust
struct S<T>(Vec<T>);
impl<T> S<T> {
    fn push(&mut self, item: T) where T: Eq { .. }
}
```

smoelius commented:
No problem at all.

Something like `S.push(&0)` would be ignored. There is a condition that the changed argument's type parameter not appear elsewhere in the function signature. In this case, `T` appears in the type of the `&mut self` argument.

smoelius commented:
However, you caused me to realize I was incorrectly assuming the argument type was a type parameter.

I have corrected the error and added a relevant test.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +931,115 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+        //   immediately.
+        // * If no applicable traits are found, `needless_borrow` remains `false`.
+        let mut needless_borrow = false;
+        for &(trait_def_id, substs) in &traits {
```
This could use `traits.iter().all(..)`. I don't see a reason to treat `Sized` specially. All the predicates must be met no matter what they are and it's already been checked that there's at least one additional predicate.

Jarcho reviewed (changes requested):
This should probably ignore cases where the trait has a `&mut self` method, but the unreferenced type is not a mutable reference. e.g.
```rust
#[derive(Clone, Copy)]
struct Iter(..);
impl Iterator for Iter { .. }

fn takes_iter(_: impl Iterator) {}
fn dont_warn(mut x: Iter) -> {
    takes_iter(&mut x);
}
fn warn(mut x: &mut Iter) -> {
    takes_iter(&mut x)
}
```

Iterators shouldn't implement `Copy` for exactly this reason, but that doesn't stop it from happening.

bors commented:
:umbrella: The latest upstream changes (presumably #9169) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

mikerite commented:
I got a crash with this branch on the crossbeam-epoch-0.9.9 crate.
```
thread 'rustc' panicked at 'assertion failed: !value.has_escaping_bound_vars()'
```

<details>
<summary>Backtrace</summary>

```
thread 'rustc' panicked at 'assertion failed: !value.has_escaping_bound_vars()', /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/compiler/rustc_middle/src/ty/sty.rs:996:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/core/src/panicking.rs:48:5
   3: <rustc_middle::ty::sty::Binder<rustc_middle::ty::sty::TraitRef>>::dummy
   4: <rustc_infer::infer::InferCtxt as rustc_trait_selection::infer::InferCtxtExt>::type_implements_trait
   5: rustc_infer::infer::InferCtxtBuilder::enter
   6: clippy_utils::ty::implements_trait_with_env
   7: clippy_utils::ty::implements_trait
   8: clippy_lints::dereference::needless_borrow_impl_arg_position
   9: <clippy_lints::dereference::Dereferencing as rustc_lint::passes::LateLintPass>::check_expr
  10: <rustc_lint::late::LateLintPassObjects as rustc_lint::passes::LateLintPass>::check_expr
  11: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  12: rustc_hir::intravisit::walk_expr::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  13: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  14: rustc_hir::intravisit::walk_expr::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  15: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  16: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_block
  17: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  18: rustc_hir::intravisit::walk_expr::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  19: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  20: rustc_hir::intravisit::walk_block::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  21: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_block
  22: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  23: rustc_hir::intravisit::walk_expr::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  24: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  25: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  26: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_block
  27: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_expr
  28: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_nested_body
  29: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_fn
  30: rustc_hir::intravisit::walk_impl_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  31: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_nested_impl_item
  32: rustc_hir::intravisit::walk_impl_item_ref::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  33: rustc_hir::intravisit::walk_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  34: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_nested_item
  35: rustc_hir::intravisit::walk_mod::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  36: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_mod
  37: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_nested_item
  38: rustc_hir::intravisit::walk_mod::<rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects>>
  39: <rustc_lint::late::LateContextAndPass<rustc_lint::late::LateLintPassObjects> as rustc_hir::intravisit::Visitor>::visit_mod
  40: rustc_lint::late::late_lint_pass_crate::<rustc_lint::late::LateLintPassObjects>
  41: rustc_lint::late::late_lint_crate::<rustc_lint::BuiltinCombinedLateLintPass>
  42: rustc_data_structures::sync::join::<rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#0}, rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#1}, (), ()>
  43: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}>
  44: <core::panic::unwind_safe::AssertUnwindSafe<rustc_interface::passes::analysis::{closure#5}::{closure#1}> as core::ops::function::FnOnce<()>>::call_once
  45: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#5}>
  46: rustc_interface::passes::analysis
  47: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), core::result::Result<(), rustc_errors::ErrorGuaranteed>>
  48: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorGuaranteed>>>
  49: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
  50: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
  51: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
  52: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  53: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.64 (2507d2ad8 2022-07-09)
```

</details>

smoelius commented:
Thanks, @mikerite!

smoelius commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +931,115 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+    cx: &LateContext<'tcx>,
+    parent: &Expr<'tcx>,
+    child_arg_index: usize,
+    child_arg_ty: Ty<'tcx>,
```
Is there an easy way to obtain a `ParamTy` here?

The `ParamTy` would replace both `child_arg_index` and `child_arg_ty` arguments, correct? If so, how would you imagine this line changing? https://github.com/rust-lang/rust-clippy/blob/2507d2ad883b56acd28506f34f4ae6918a32173c/clippy_lints/src/dereference.rs#L996

Jarcho replied:
It would only replace `child_arg_ty`. Would look like
```rust
if let ty::Param(ty) = ty {
    needless_borrow_impl_arg_position(...)
} else {
    param_auto_deref_stability(...)
}
```
You would have to change `needless_borrow_impl_arg_position` to return `Position::Other` which would be correct for any generic param.

smoelius replied:
Are there no cases where `ty` is a `Param` and `needless_borrow_impl_arg_position` would return `None` but `param_auto_deref_stability` returns `Some(..)`? Put another way, are we sure it is okay to elide the current code's `unwrap_or_else`?

Separately, could you please explain why this change necessitates the use of `Position::Other`? I am currently using `Position::ImplArg` to determine the error message: https://github.com/rust-lang/rust-clippy/blob/2507d2ad883b56acd28506f34f4ae6918a32173c/clippy_lints/src/dereference.rs#L340-L341

Jarcho replied:
`param_auto_deref_stability` returns `Position::Other(_)` for generic types.

There's no need to return an `Option` from `needless_borrow_impl_arg_position `. Either you can remove borrows, or nothing interesting can be done with the location the expression is in. The latter case uses `Position::Other(_)`.

smoelius commented on clippy_lints/src/dereference.rs:
```diff
@@ -751,12 +758,14 @@ fn walk_parents<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) -> (Position, &
+                            // Type inference for closures can depend on how they're called. Only go by the explicit
+                            // types here.
+                            Some(ty) => binding_ty_auto_deref_stability(ty, precedence),
+                            None => needless_borrow_impl_arg_position(cx, parent, i, ty.skip_binder(), child_id)
```
Is the idea that, when `needless_borrow_impl_arg_position` applies, `child_id` should refer to the same expression as the one passed into `walk_parents`? (I'm trying to understand why it seems to work now.)

smoelius commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +931,115 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+    let check_referent = |referent| {
+        let referent_ty = cx.typeck_results().expr_ty(referent);
+
+        if !is_copy(cx, referent_ty) {
```
What would you recommend here? I looked to [`trivially_copy_pass_by_ref`](https://rust-lang.github.io/rust-clippy/master/#trivially_copy_pass_by_ref) for inspiration, but it seems to have no size limit by default.

Jarcho replied:
It's limited to 8 bytes by default. A little extreme as a fat pointer should still be passed by value in general. Can you check through all the types that were caught in clippy and see what the sizes are?

smoelius replied:
It looks like this was the largest at 160 bytes: https://github.com/rust-lang/rust-clippy/blob/2507d2ad883b56acd28506f34f4ae6918a32173c/tests/lint_message_convention.rs#L20-L31

smoelius commented:
I think the most recent push addresses all of the review comments up to this point, and the crash. Please let me know if I misunderstood anything.

Regarding the "is copyable" condition, I imposed a hard limit of 256 bytes. Please let me know if you would prefer a different approach.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +943,148 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+    // If no traits were found, or only the `Sized` or `Any` traits were found, return.
+    if trait_predicates.iter().all(|trait_predicate| {
+        let trait_def_id = trait_predicate.def_id();
+        Some(trait_def_id) == sized_trait_def_id || match_def_path(cx, trait_def_id, &paths::ANY_TRAIT)
```
Should use `cx.tcx.is_diagnostic_item(sym::Any, trait_def_id)`

Jarcho commented:
A couple of comments now that I've had time to think through the implementation.
* You can ignore my earlier comment about the size of parameters. The calling convention for large types is that they're passed by reference anyways. Sorry for wasting time on that one.
* Currently this only handles `T: SomeTrait` predicates, but `T: SomeTrait<U>` and `U: SomeTrait<T>` predicates also need to be handled. This requires a few steps:
    1. Get the current substitution list for the function call.
    2. Replace the substitution for `T` with the new type.
    3. Apply the new substitution list to each predicate.
    4. Check each new substituted predicate with [`predicate_must_hold_modulo_regions`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/query/evaluate_obligation/trait.InferCtxtExt.html#tymethod.predicate_must_hold_modulo_regions). The `InferCtxt` can be created with `cx.tcx.infer_ctxt().enter(|infcx| ...)`

  You'll still want to keep the rest of the restrictions in place in order to avoid false positives.

smoelius commented:
> * Currently this only handles `T: SomeTrait` predicates, but `T: SomeTrait<U>` and `U: SomeTrait<T>` predicates also need to be handled. This requires a few steps:
>   
>   1. Get the current substitution list for the function call.
>   2. Replace the substitution for `T` with the new type.
>   3. Apply the new substitution list to each predicate.
>   4. Check each new substituted predicate with [`predicate_must_hold_modulo_regions`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/query/evaluate_obligation/trait.InferCtxtExt.html#tymethod.predicate_must_hold_modulo_regions). The `InferCtxt` can be created with `cx.tcx.infer_ctxt().enter(|infcx| ...)`

You're saying this should be done for all predicates, not just trait predicates, correct?

If so, are there any tools you know of to make iii easier? I.e., something easier than performing a match and handling each `PredicateKind` individually?

Jarcho commented:
The [`subst`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/subst/trait.Subst.html#tymethod.subst) method will perform the actual substitution. It exists on both `List<Predicate>>` and `Predicate` itself

bors commented:
:umbrella: The latest upstream changes (presumably #9210) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

smoelius commented:
I think the second round of comments are addressed.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +941,177 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+            return false;
+        }
+
+        let substs_with_expr_ty = substs_with_expr_ty.split_at(callee_generics.parent_count);
```
This whole thing can be replaced with `substs[param_ty.index as usize] = ty::GenericArg::from(referent_ty);` assuming `substs` is a slice containing the original substitutions.

smoelius replied:
Wow. That is a lot simpler.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +941,177 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+            .collect::<Vec<_>>();
+
+        let predicates = EarlyBinder(predicates).subst(cx.tcx, &substs_with_referent_ty);
+        if !predicates.iter().all(|predicate| {
```
The result of `all` can just be returned directly.

smoelius commented:
Please let me know if I misunderstood any of your suggestions, @Jarcho.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +941,150 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+
+        let mut substs_with_referent_ty = substs_with_expr_ty.to_vec();
+        substs_with_referent_ty[param_ty.index as usize] = ty::GenericArg::from(referent_ty);
+        for projection_predicate in &projection_predicates {
```
This loop can just be removed. It takes any predicate `<T as Trait>::Assoc = U` and modifies the substitution list replacing `U` with `<V as Trait>::Assoc` where `V` is the argument type minus the reference, and `U` is some type parameter.

The only item in the substitution list that needs to change is the one for the argument being modified.

smoelius replied:
Without the loop on line 1035, this test fails: https://github.com/rust-lang/rust-clippy/blob/cd70a3d3b1c8e019c8938b259bac496b4179dd75/tests/ui/needless_borrow.rs#L131
In particular, `substs_with_referent_ty` looks like this:
```
[[&str; 2], &&str]
```
And the following predicate fails:
```
Binder(ProjectionPredicate(ProjectionTy { substs: [[&str; 2]], item_def_id: DefId(2:8368 ~ core[32c2]::iter::traits::collect::IntoIterator::Item) }, Ty(&&str)), [])
```
With the loop, `substs_with_referent_ty` looks like this:
```
[[&str; 2], &str]
```
Essentially, the types projected from `&T` and `T` differ.

I am open to doing this differently. What I have now feels manual and is certainly not general.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +941,150 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+            return false;
+        }
+
+        let mut substs_with_referent_ty = substs_with_expr_ty.to_vec();
```
The `Vec` can be constructed outside the closure. The same item is being modified in every call.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +941,150 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+        .iter()
+        .filter_map(|predicate| {
+            if let PredicateKind::Trait(trait_predicate) = predicate.kind().skip_binder()
+                && is_substs_for_type(trait_predicate.trait_ref.substs, param_ty.to_ty(cx.tcx))
```
This could be `trait_predicate.trait_ref.self_ty() == param_ty.to_ty(cx.tcx)`

smoelius commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +941,150 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+        });
+
+    let predicates = cx.tcx.param_env(callee_def_id).caller_bounds();
+    let trait_predicates = predicates
```
This vector is also used in the `has_ref_mut_self_method` check, actually.

Jarcho replied:
Give the current code all the checks could be combined into a single `any` call.

smoelius commented:
@Jarcho How's this?

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +944,191 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+        }
+
+        // The next line should not panic, but panicking seems better than going into an infinite loop.
+        assert!(replaced.insert(param_ty.index));
```
I believe this panics on:
```rust
trait X {
    type X;
}
impl X for () {
    type X = ();
}

fn f<T, U, Z>(_: T)
where T: X<X = U>,
    U: X<X = Z>,
    T: X<X = Z>,
{}

fn main() {
    f(());
}
```

I'm not sure why anyone would type this, but you can safely continue to the next type in this case.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +944,191 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+        let new_generic_arg = ty::GenericArg::from(new_ty);
+
+        // If `replaced.is_empty()`, then `param_ty` and `new_ty` are those initially passed in.
+        if !(substs[param_ty.index as usize] == new_generic_arg
```
This check could be done right before pushing the types to the queue thereby removing the need to check for the initial argument.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +944,191 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+                    .mk_projection(assoc_item.def_id, cx.tcx.mk_substs_trait(new_ty, &[]));
+                let projected_ty = cx.tcx.normalize_erasing_regions(cx.param_env, projection);
+
+                deque.push_back((*term_param_ty, projected_ty));
```
Combined with the previous change this could modify the substitution for `term_param_ty` immediately rather than storing it for later. The check for `replaced.insert(...)` could also be moved here before pushing the item onto the queue.

smoelius commented:
Is this about what you had in mind? If so, I am concerned that the optimizations are starting to affect the code's maintainability.

I personally do not like making the `any`'s closure not `Fn`.

And while unrolling loop eliminates a check, it also causes some code to be duplicated.

I don't have a strong opinion, but I do think the code was more clear with the allocation and the loop rolled.

Apologies if I have misunderstood your suggestions.

Jarcho commented:
You could use `inspect` to separate the two, or just use a for loop.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +944,201 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+    let mut replaced = BitSet::new_empty(substs.len());
+    let mut deque = VecDeque::with_capacity(substs.len());
+
+    if fn_sig
```
I had missed that you removed the check from before `replace_types`. I had thought you still had it near the beginning of `needless_borrow_impl_arg_position`. You can revert the loop unrolling.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +944,201 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+        .inputs_and_output
+        .iter()
+        .enumerate()
+        .any(|(i, ty)| i != arg_index && contains_ty(ty, param_ty.to_ty(cx.tcx)))
```
I had missed that you removed the check from before `replace_types`. I had thought you still had it near the beginning of `needless_borrow_impl_arg_position`. You can revert the loop unrolling.

Jarcho commented on clippy_lints/src/dereference.rs:
```diff
@@ -920,6 +944,201 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+                }
+
+                // The next line provides some protection against infinite loops.
+                assert!(inserted);
```
This can still trigger, though the case is much more contrived and requires explicitly passing a generic argument. You can do:
```rust
if insert {
    // All the work in here
}
```
The only way you can fail to insert is if the type has already been placed in the queue. If this happens the inference has already been propagated, or it will be in a future iteration. In both cases nothing needs to happen.

smoelius commented:
I reverted the loop unrolling and made it so that `has_ref_mut_self_method` is called from a closure called by `inspect`.

smoelius commented:
> * `Some("").unwrap_or(&"")` gives the new error message when it should give the original one.

What would make sense here? Ignore `core::marker::Destruct`, like we do `Any` and `Sized`? https://doc.rust-lang.org/src/core/option.rs.html#776-778

Jarcho commented:
For this case `T` appears three times in the signature, so extra bounds shouldn't matter. `Destruct` should probably be ignored as well.

smoelius commented:
> For this case `T` appears three times in the signature, so extra bounds shouldn't matter.

For the example of `Some("").unwrap_or(&"")`, the substitution for `T` in `unwrap_or` is `&str` like in `Option::Some`. So the left side of this `&&` fails, avoiding the `return false`: https://github.com/rust-lang/rust-clippy/blob/5c5bf96be958a72effd9cce4f6379823bdbe93a2/clippy_lints/src/dereference.rs#L1098-L1103
(Admittedly, I am surprised this example compiles at all.)

However, ignoring `Destruct` produces the original error message, as desired.

Please pay particular attention to how I added the MSRV checks, as this is my first attempt at it.

Jarcho commented:
There's a type inference hack to make it adjustments work on types inferred earlier. It doesn't compile if you flip the where the reference is.

Is that first condition necessary? As far as I can tell it should only happen in this case which will be picked up by the normal logic, and if the type has already been modified in an earlier iteration. The early escape on the iteration can be handled by putting everything within the `replaced.insert` check.

The msrv check looks good. 👍

smoelius commented:
> Is that first condition necessary?

The first condition was mean to address https://github.com/rust-lang/rust-clippy/pull/9136#discussion_r927212757:
> * `U` doesn't exist in the argument list or return position.

> ...you can still lint, but only if the projection type normalizes to the same type for both the original type and the replacement type.

But maybe I'm not thinking about this right.

Jarcho commented:
You would need to use [`try_normalize_erasing_regions`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.try_normalize_erasing_regions) to do that. Two projection types from different bases are never considered equal, but the can normalize to the same type.

I think the conditions you want on the loop would be:
* `replaced.insert(..) -> continue`: The type has already been replaced so there's no need to do anything.
* `original_ty == replacement_ty -> continue`: The type is the same so there's nothing to change.
* `sig_contains_ty_param -> return false`: One of the types can't be replaced.

Jarcho commented:
I think that should work. The test would be calling a function like `fn f<T, U>(_: T, _: U) where T: Trait<Assoc=U> {}`. Both `&T` and `T` would need to have the same type for `Assoc`.

smoelius commented:
What you're describing sounds very similar to this test: https://github.com/rust-lang/rust-clippy/blob/7e18f4fd9b1377ad95deda49f60f31ecc3ccdd54/tests/ui/needless_borrow.rs#L135

But I think I am losing the thread. What behavior do you expect when calling a function like that?

Jarcho commented:
I really should have waited for when I had more time to review this. Totally missed that you were already normalizing the projection when making that last comment.

smoelius commented:
No problem. But do you think the check was better in its previous position?

Jarcho commented:
The current position is fine. It also avoids the same reading problem I made.

I think that needs to be `try_normalize_erasing_regions`. You don't have a case where `&T` implements a trait with a projection, but `T` doesn't. I think the `normalize_erasing_regions` ICE's in that case.

smoelius commented:
Please consider this paused. I have to fix this.

smoelius commented:
OK. I think this is fixed now. The new test ICEs without `try_normalize_erasing_regions`.

Jarcho commented:
That should be the last thing. I want to run one more check on more crates, but that will take a couple of days as I have other tests running right now.

Jarcho commented:
Looks like it mostly catches `AsRef` and `IntoIterator` with `Serialize` being a very distant third. Didn't see any FP's in the top 5000 crates.

You can squash down some of the commits and then it's good to go.

smoelius commented:
Done.

I left in the change that moved `get_input_traits_and_projections` to `clippy_utils`, even though this change was ultimately not needed. Please let me know if you would like me to undo this change.

Jarcho commented:
It would be better to remove it. This would get removed as an unused util function if we ever clean up useless/overlapping util functions.

I'm going to put a hold on merging this as it conflicts #9126 and the lint fixed there is about to beta. I can't guarantee I'll have time to deal with the merge conflict in the next week.

smoelius commented:
> It would be better to remove it. This would get removed as an unused util function if we ever clean up useless/overlapping util functions.

I moved `get_input_traits_and_projections` back into `unnecessary_to_owned.rs`, but kept a few changes that came out of this review.

> I'm going to put a hold on merging this as it conflicts #9126 and the lint fixed there is about to beta. I can't guarantee I'll have time to deal with the merge conflict in the next week.

No problem at all.

Jarcho commented:
Ping @smoelius just in case github didn't. This is good to merge once the conflicts are resolved. Sorry for dumping that on you.

smoelius commented:
Sorry, I misunderstood your previous message.

If it helps to verify the rebase, here is the diff between the old patch (ad778ad) and the new (032f112):
```rust
15c15
< index d513a229b..47bdc41a2 100644
---
> index 3b27f061e..357cf6fc4 100644
25c25
<              return Err(CliError::RaSetupActive);
---
>              return Err(CliError::IntellijSetupActive);
50c50
< index c089f4d8c..c831fb831 100644
---
> index 05e79a241..c503142e5 100644
72c72
< index a90f894a7..7d2b526bc 100644
---
> index e0b94f719..e38704701 100644
79c79
< -use clippy_utils::ty::{expr_sig, peel_mid_ty_refs, variant_of_res};
---
> -use clippy_utils::ty::{expr_sig, peel_mid_ty_refs, ty_sig, variant_of_res};
81c81
< +use clippy_utils::ty::{contains_ty, expr_sig, is_copy, peel_mid_ty_refs, variant_of_res};
---
> +use clippy_utils::ty::{contains_ty, expr_sig, is_copy, peel_mid_ty_refs, ty_sig, variant_of_res};
99c99
< -use rustc_middle::ty::{self, Ty, TyCtxt, TypeVisitable, TypeckResults};
---
> -use rustc_middle::ty::{self, Binder, BoundVariableKind, List, Ty, TyCtxt, TypeVisitable, TypeckResults};
101,102c101,102
< +    self, subst::Subst, EarlyBinder, FnSig, GenericArgKind, ParamTy, PredicateKind, ProjectionPredicate, Ty, TyCtxt,
< +    TypeVisitable, TypeckResults,
---
> +    self, subst::Subst, Binder, BoundVariableKind, EarlyBinder, FnSig, GenericArgKind, List, ParamTy, PredicateKind,
> +    ProjectionPredicate, Ty, TyCtxt, TypeVisitable, TypeckResults,
106c106
<  use rustc_span::{symbol::sym, Span, Symbol};
---
>  use rustc_span::{symbol::sym, Span, Symbol, DUMMY_SP};
196c196
< @@ -521,7 +547,7 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing {
---
> @@ -510,7 +536,7 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing {
205c205
< @@ -553,6 +579,8 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing {
---
> @@ -542,6 +568,8 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing {
214c214
< @@ -605,6 +633,7 @@ enum Position {
---
> @@ -594,6 +622,7 @@ enum Position {
222c222
< @@ -638,7 +667,7 @@ impl Position {
---
> @@ -630,7 +659,7 @@ impl Position {
228c228
<              Self::DerefStable(p) | Self::ReborrowStable(p) | Self::Other(p) => p,
---
>              Self::DerefStable(p, _) | Self::ReborrowStable(p) | Self::Other(p) => p,
231c231
< @@ -647,8 +676,12 @@ impl Position {
---
> @@ -639,8 +668,12 @@ impl Position {
246c246
< @@ -751,12 +784,19 @@ fn walk_parents<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) -> (Position, &
---
> @@ -732,13 +765,20 @@ fn walk_parents<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) -> (Position, &
254,255c254,256
< -                        Some(ty) => binding_ty_auto_deref_stability(ty, precedence),
< -                        None => param_auto_deref_stability(ty.skip_binder(), precedence),
---
> -                        Some(hir_ty) => binding_ty_auto_deref_stability(cx, hir_ty, precedence, ty.bound_vars()),
> -                        None => ty_auto_deref_stability(cx, cx.tcx.erase_late_bound_regions(ty), precedence)
> -                            .position_for_arg(),
260c261
< +                            Some(ty) => binding_ty_auto_deref_stability(ty, precedence),
---
> +                            Some(hir_ty) => binding_ty_auto_deref_stability(cx, hir_ty, precedence, ty.bound_vars()),
265c266,267
< +                                    param_auto_deref_stability(ty.skip_binder(), precedence)
---
> +                                    ty_auto_deref_stability(cx, cx.tcx.erase_late_bound_regions(ty), precedence)
> +                                        .position_for_arg()
272c274
< @@ -797,7 +837,12 @@ fn walk_parents<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) -> (Position, &
---
> @@ -779,12 +819,17 @@ fn walk_parents<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) -> (Position, &
276c278,283
< -                            param_auto_deref_stability(cx.tcx.fn_sig(id).skip_binder().inputs()[i], precedence)
---
> -                            ty_auto_deref_stability(
> -                                cx,
> -                                cx.tcx.erase_late_bound_regions(cx.tcx.fn_sig(id).input(i)),
> -                                precedence,
> -                            )
> -                            .position_for_arg()
281c288,293
< +                                param_auto_deref_stability(ty, precedence)
---
> +                                ty_auto_deref_stability(
> +                                    cx,
> +                                    cx.tcx.erase_late_bound_regions(cx.tcx.fn_sig(id).input(i)),
> +                                    precedence,
> +                                )
> +                                .position_for_arg()
286c298
< @@ -920,6 +965,205 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
---
> @@ -946,6 +991,205 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
489,492c501,504
<  // Checks whether a type is stable when switching to auto dereferencing,
<  fn param_auto_deref_stability(ty: Ty<'_>, precedence: i8) -> Position {
<      let ty::Ref(_, mut ty, _) = *ty.kind() else {
< @@ -1026,7 +1270,8 @@ fn report<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, state: State, data
---
>  struct TyPosition<'tcx> {
>      position: Position,
>      ty: Option<Ty<'tcx>>,
> @@ -1084,7 +1328,8 @@ fn report<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, state: State, data
503c515
< index 8444532f6..ef8c33a03 100644
---
> index e6a405f81..521739c28 100644
506c518
< @@ -807,7 +807,7 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
---
> @@ -821,7 +821,7 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
516c528
< index b4c6bfb31..ad55ee63c 100644
---
> index 99b56da7a..e4624167a 100644
519,521c531,533
< @@ -5,9 +5,8 @@ use clippy_utils::source::snippet_opt;
<  use clippy_utils::ty::{
<      contains_ty, get_associated_type, get_iterator_item_ty, implements_trait, is_copy, peel_mid_ty_refs,
---
> @@ -6,9 +6,8 @@ use clippy_utils::ty::{
>      contains_ty, get_associated_type, get_iterator_item_ty, implements_trait, is_copy, is_type_diagnostic_item,
>      peel_mid_ty_refs,
530c542
< @@ -360,25 +359,15 @@ fn get_input_traits_and_projections<'tcx>(
---
> @@ -373,25 +372,15 @@ fn get_input_traits_and_projections<'tcx>(
```

Jarcho commented:
Thank you for pushing through and implementing all this.

@bors r+

bors commented:
:pushpin: Commit 032f112745d2632fc37175fad7bab98215062586 has been approved by `Jarcho`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Jarcho 032f112745d2632fc37175fad7bab98215062586 -->
<!-- homu: {"type":"Approved","sha":"032f112745d2632fc37175fad7bab98215062586","approver":"Jarcho","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 032f112745d2632fc37175fad7bab98215062586 with merge c419d0a8b538de6000226cc54a2f18a03bbd31d6...
<!-- homu: {"type":"BuildStarted","head_sha":"032f112745d2632fc37175fad7bab98215062586","merge_sha":"c419d0a8b538de6000226cc54a2f18a03bbd31d6"} -->

smoelius commented:
Thanks for the detailed and thorough review!

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/7902528496?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/7902508116?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/7902782784?check_suite_focus=true)
Approved by: Jarcho
Pushing c419d0a8b538de6000226cc54a2f18a03bbd31d6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Jarcho","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/7902782784?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/7902528496?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/7902508116?check_suite_focus=true"},"merge_sha":"c419d0a8b538de6000226cc54a2f18a03bbd31d6"} -->
